src/wallet.rs

```diff
@@ -0,0 +1,320 @@
+///
+/// # Example
+///
+/// #[ignore]
+/// ```
+/// use mobile_sdk_rs::prelude::*;
+///
+/// let storage_manager = Box::new(MyStorageManager::new());
+/// let key_manager = Box::new(MyKeyManager::new());
+///
+/// fn main() -> Result<(), WalletError> {
+///  let wallet = Wallet::new(storage_manager, key_manager)?;
+///
+/// }
+///
+///
+/// ```
+#[derive(uniffi::Object, Debug)]
+pub struct Wallet {
+    pub(crate) client: oid4vp::core::util::ReqwestClient,
+    pub(crate) metadata: MetadataManager,
+    pub(crate) vdc_collection: VdcCollection,
+    // NOTE: The wallet has internal access to the trust manager APIs, but
+    // is currently not exposing them to the foreign function interface.
+    // This is because the trust manager is intended to be used internally by the wallet.
+    // The [TrustManager] does not implement uniffi bindings.
+    pub(crate) trust_manager: TrustManager,
+    pub(crate) key_manager: Box<dyn KeyManagerInterface>,
+    pub(crate) storage_manager: Box<dyn StorageManagerInterface>,
+    // The active key index is used to determine which key to used for signing.
+    // By default, this is set to the 0-index key.
+    pub(crate) active_key_index: Arc<RwLock<u8>>,
+}
+
+#[uniffi::export]
+impl Wallet {
+    /// Initialize a new [Wallet] instance provides a storage manager and key manager.
+    ///
+    /// The storage and key managers are foreign callback interfaces that are implemented
+    /// in the native language of the foreign environment, e.g. Kotlin, Swift, etc.
+    ///
+    /// # Arguments
+    ///
+    /// * `storage_manager` - A foreign callback interface that implements the [StorageManagerInterface].
+    ///
+    /// * `key_manager` - A foreign callback interface that implements the [KeyManagerInterface].
+    ///
+    /// # Returns
+    ///
+    /// A new [Wallet] instance.
+    ///
+    /// # Errors
+    ///
+    /// * If the HTTP client fails to initialize;
+    /// * If there is a storage error when initializing the metadata manager.
+    #[uniffi::constructor]
+    pub fn new(
+        storage_manager: Box<dyn StorageManagerInterface>,
+        key_manager: Box<dyn KeyManagerInterface>,
+    ) -> Result<Arc<Self>, WalletError> {
+        let client = oid4vp::core::util::ReqwestClient::new()
+            .map_err(|e| WalletError::HttpClientInitialization(format!("{e}")))?;
+
+        // Initalized the metdata manager with the device storage.
+        // This will load any existing metadata from storage, otherwise
+        // it will create a new metadata instance.
+        let metadata = MetadataManager::initialize(&storage_manager)?;
+
+        Ok(Arc::new(Self {
+            client,
+            metadata,
+            key_manager,
+            storage_manager,
+            vdc_collection: VdcCollection::new(),
+            trust_manager: TrustManager::new(),
+            active_key_index: Arc::new(RwLock::new(DEFAULT_KEY_INDEX)),
```

Comment:
Consider creating a `KeyRing` struct to maintain the active key index and ensure that the device has at least one signing key available. Creating a default key if one does not exist.

Reply:
Not needed here.